src/dfx/src/commands/generate.rs

```diff
@@ -31,40 +33,62 @@ pub fn exec(env: &dyn Environment, opts: GenerateOpts) -> DfxResult {
     env.get_cache().install()?;
 
     // Option can be None which means generate types for all canisters
-    let canister_names = config
+    let canisters_to_load = config
         .get_config()
         .get_canister_names_with_dependencies(opts.canister_name.as_deref())?;
+    let canisters_to_generate = canisters_to_load
+        .clone()
+        .into_iter()
+        .filter(|canister_name| {
+            !matches!(
+                config
+                    .get_config()
+                    .get_remote_canister_id(canister_name, &env.get_network_descriptor().name),
+                Ok(Some(_))
+            )
+        })
+        .collect();
 
-    // Get pool of canisters to build
-    let canister_pool = CanisterPool::load(&env, false, &canister_names)?;
+    let canister_pool = CanisterPool::load(&env, false, &canisters_to_load)?;
 
-    // This is just to display an error if trying to generate before creating the canister.
+    // This is just to display an error if trying to generate before creating the canister(s).
     let store = CanisterIdStore::for_env(&env)?;
+
     // If generate for motoko canister, build first
-    let mut build_before_generate = false;
+    let mut build_before_generate = Vec::new();
     for canister in canister_pool.get_canister_list() {
         let canister_name = canister.get_name();
         let canister_id = store.get(canister_name)?;
         if let Some(info) = canister_pool.get_canister_info(&canister_id) {
             if info.is_motoko() {
-                build_before_generate = true;
+                build_before_generate.push(canister_name.to_string());
+                trace!(
+                    log,
+                    "Found Motoko canister '{}' - will have to build before generating IDL.",
+                    canister_name
+                );
             }
         }
     }
 
-    let build_config = BuildConfig::from_config(&config)?;
+    let build_config =
+        BuildConfig::from_config(&config)?.with_canisters_to_build(build_before_generate);
+    let generate_config =
+        BuildConfig::from_config(&config)?.with_canisters_to_build(canisters_to_generate);
 
-    if build_before_generate {
-        slog::info!(
-            env.get_logger(),
-            "Building canisters before generate for Motoko"
-        );
+    if !build_config
+        .canisters_to_build
+        .as_ref()
+        .map(|v| v.is_empty())
+        .unwrap_or(true)
```

Comment:
I had some trouble understanding which direction this was going, in particular the logic around 'what if it's empty / what if it's None)', especially with respect to `.unwrap_or(true)`.  What would you think about this instead?
```suggestion
    if build_config
        .canisters_to_build
        .as_ref()
        .map(|v| !v.is_empty())
        .unwrap_or(false)
```

Reply:
much better!